File: 126-word-ladder-ii/126-word-ladder-ii.py
import copy
class Solution(object):
    def findLadders(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: List[List[str]]
        """
        #build the graph
        modified_to_word = defaultdict(list)
        word_to_modified = defaultdict(list)
        for word in wordList:
            for i, char in enumerate(word):
                word_modified_by_one = word[:i] + '*' + word[i+1:]
                modified_to_word[word_modified_by_one].append(word)
                word_to_modified[word].append(word_modified_by_one)
        if endWord not in word_to_modified:
            return []        
        graph = defaultdict(set)
        if beginWord not in word_to_modified:
            for i, char in enumerate(beginWord):
                word_modified_by_one = beginWord[:i] + '*' + beginWord[i+1:]
                modified_to_word[word_modified_by_one].append(beginWord)
                word_to_modified[beginWord].append(word_modified_by_one)
            wordList.append(beginWord)
        for word in wordList:
            for modified in word_to_modified[word]:
                for nxt in modified_to_word[modified]:
                    graph[word].add(nxt)


        #Dijkstra
        distance = defaultdict(int)
        prev_nodes = defaultdict(set)
        visited = set()
        step = 0
        queue = deque()
        queue.append(beginWord)
        visited.add(beginWord)
        while(queue):
            size = len(queue)
            step += 1
            for i in range(size):
                popped = queue.popleft()
                
                for nxt in graph[popped]:
                    if nxt in visited:
                        if step == distance[nxt]:
                            prev_nodes[nxt].add(popped)
                        continue
                    visited.add(nxt)
                    distance[nxt] = step
                    queue.append(nxt)
                    prev_nodes[nxt].add(popped)
        
        #search from endWord
        queue = deque()
        queue.append([endWord])
        res = []
        while(queue):
            size = len(queue)
            
            for i in range(size):
                popped = queue.popleft()
                last_word = popped[-1]
                if last_word == beginWord:
                    popped_reverse = popped[::-1]
                    res.append(popped_reverse)
                    continue
                for prev in prev_nodes[last_word]:
                    new_element = popped[:]
                    new_element.append(prev)
                    queue.append(new_element)
            if res:
                return res
        return res
        
            
        

File: 126-word-ladder-ii/126-word-ladder-ii.java
class Solution {
    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        //build the graph
        
        HashMap<String, Set<String>> graph = new HashMap<>();
        
        boolean flag = buildGraph(beginWord, endWord, wordList, graph);
        if(!flag)return new LinkedList<>();
        
        LinkedList<LinkedList<String>> queue = new LinkedList<>();
        
        Map<String, Set<String>> prevNodes = new HashMap<>();
        buildPrevNodes(graph, beginWord, prevNodes);
        return BFS( beginWord, endWord, prevNodes, graph);
    
    }
    public List<List<String>> BFS(String beginWord, String endWord, Map<String, Set<String>> prevNodes, Map<String, Set<String>> graph){
        LinkedList<LinkedList<String>> queue = new LinkedList<>();
        LinkedList<List<String>> res = new LinkedList<>();
        queue.add(new LinkedList<>(Arrays.asList(endWord)));
        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i = 0; i < size; i++){
                LinkedList<String> tempRes = queue.pollFirst();
                String tempWord = tempRes.get(tempRes.size() - 1);
                if(tempWord.equals(beginWord)){
                    List<String> added = (List) tempRes.clone();
                    Collections.reverse(added);
                    res.add(added);
                    continue;
                } 
            if(!prevNodes.containsKey(tempWord))continue;
            for(String nxt : prevNodes.get(tempWord)){
                    LinkedList<String> added = (LinkedList) tempRes.clone();
                    added.add(nxt);
                    queue.add(added);
            
            }
        
            }
            if(!res.isEmpty())return res;
        
        }
        
        return res;
        
    }
    
    
    
    public void buildPrevNodes(Map<String, Set<String>> graph, String beginWord, Map<String, Set<String>> prevNodes){
        LinkedList<String> queue = new LinkedList<>();
        Map<String, Integer> distance = new HashMap<>();
        Set<String> visited = new HashSet<>();
        int step = 0;
        queue.add(beginWord);
        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i = 0; i < size; i++){
            String popped = queue.pollFirst();
            distance.put(popped, step);
            for(String modified : modifyString(popped)){
                if(!graph.containsKey(modified))continue;
                for(String nxt : graph.get(modified)){
                    if(distance.containsKey(nxt))continue;
                    if(!visited.contains(nxt)){
                        queue.addLast(nxt);
                        visited.add(nxt);
                    }
                    prevNodes.computeIfAbsent(nxt, k->new HashSet<>());
                    prevNodes.get(nxt).add(popped);
                }
            }
            }
            step += 1;
        }
    }
    public String[] modifyString(String original){
        int n = original.length();
        String[] res = new String[n];
        for(int i = 0; i < n; i++){
            StringBuilder target = new StringBuilder();
            target.append(original.substring(0, i));
            target.append("*");
            target.append(original.substring(i+1));
            res[i] = target.toString();
        }
        return res;
    }
    public boolean buildGraph(String beginWord, String endWord, List<String> wordList, HashMap<String, Set<String>> graph){
        wordList.add(beginWord);
        boolean flag = false;
        for(String word : wordList){
            if(word.equals(endWord)){
                flag = true;
            }
            String[] modifiedStrings = modifyString(word);
            for(String modifiedString : modifiedStrings){
                graph.computeIfAbsent(modifiedString, k -> new HashSet<String>());
                graph.get(modifiedString).add(word);
            }
        }
        return flag;
        
    }
    
}